From 0c5b99ebaf708f9917f2271781bc94c53150dcd5 Mon Sep 17 00:00:00 2001
From: skyjet18 <57456827+skyjet18@users.noreply.github.com>
Date: Tue, 4 Mar 2025 19:15:53 +0100
Subject: [PATCH 29/33] hls: improved handling of event type playlist

- allow seeking in event playlist
- event playlist will be started from beggining instead of end
- event playlist is refreshed the same way as live type playlist
---
 libavformat/hls.c | 29 ++++++++++++++++++-----------
 1 file changed, 18 insertions(+), 11 deletions(-)

diff --git a/libavformat/hls.c b/libavformat/hls.c
index b1349a7de4..c778c629fb 100644
--- a/libavformat/hls.c
+++ b/libavformat/hls.c
@@ -315,6 +315,20 @@ static void free_rendition_list(HLSContext *c)
     c->n_renditions = 0;
 }
 
+static void update_duration(AVFormatContext *s)
+{
+    HLSContext *c = s->priv_data;
+    int i;
+
+     /* If this isn't a live stream, calculate the total duration of the stream. */
+    if (c->n_variants > 0 && (c->variants[0]->playlists[0]->finished || c->variants[0]->playlists[0]->type == PLS_TYPE_EVENT)) {
+        int64_t duration = 0;
+        for (i = 0; i < c->variants[0]->playlists[0]->n_segments; i++)
+            duration += c->variants[0]->playlists[0]->segments[i]->duration;
+        s->duration = duration;
+    }
+}
+
 static struct playlist *new_playlist(HLSContext *c, const char *url,
                                      const char *base)
 {
@@ -1046,6 +1060,8 @@ fail:
         !(c->variants[0]->playlists[0]->finished ||
           c->variants[0]->playlists[0]->type == PLS_TYPE_EVENT))
         c->ctx->ctx_flags |= AVFMTCTX_UNSEEKABLE;
+
+    update_duration(c->ctx);
     return ret;
 }
 
@@ -1892,12 +1908,12 @@ static int64_t select_cur_seq_no(HLSContext *c, struct playlist *pls)
     /* If playback is already in progress (we are just selecting a new
      * playlist) and this is a complete file, find the matching segment
      * by counting durations. */
-    if (pls->finished && c->cur_timestamp != AV_NOPTS_VALUE) {
+    if ((pls->finished || pls->type == PLS_TYPE_EVENT) && c->cur_timestamp != AV_NOPTS_VALUE) {
         find_timestamp_in_playlist(c, pls, c->cur_timestamp, &seq_no, NULL);
         return seq_no;
     }
 
-    if (!pls->finished) {
+    if (!pls->finished && pls->type != PLS_TYPE_EVENT) {
         if (!c->first_packet && /* we are doing a segment selection during playback */
             c->cur_seq_no >= pls->start_seq_no &&
             c->cur_seq_no < pls->start_seq_no + pls->n_segments)
@@ -2120,15 +2136,6 @@ static int hls_read_header(AVFormatContext *s)
         }
     }
 
-    /* If this isn't a live stream, calculate the total duration of the
-     * stream. */
-    if (c->variants[0]->playlists[0]->finished) {
-        int64_t duration = 0;
-        for (i = 0; i < c->variants[0]->playlists[0]->n_segments; i++)
-            duration += c->variants[0]->playlists[0]->segments[i]->duration;
-        s->duration = duration;
-    }
-
     /* Associate renditions with variants */
     for (i = 0; i < c->n_variants; i++) {
         struct variant *var = c->variants[i];
-- 
2.45.2.windows.1

